### ISSUE TITLE
Complex nested table recognition failure in financial documents with small fonts

### ISSUE THREAD

**Issue by zhixiangxue** [deepseek-ai/DeepSeek-OCR-2#12]
# Complex nested table recognition failure in financial documents with small fonts

## Environment
- **Model**: DeepSeek-OCR-2 (Transformers)
- **Version**: Latest from HuggingFace Hub (deepseek-ai/DeepSeek-OCR-2)
- **Hardware**: RTX 4090 (24GB VRAM)
- **Software**: 
  - Python 3.12.9
  - PyTorch 2.6.0+cu118
  - Transformers 4.46.3
  - Windows 11

## Issue Description

The model performs excellently on simple tables but fails significantly on **complex nested financial tables with small fonts** (6-7pt). Specifically, the "PROGRAM HIGHLIGHTS" section in mortgage rate sheets shows ~15% accuracy compared to 90%+ for simpler tables in the same document.

## Test Image

<img width="701" height="876" alt="Image" src="https://github.com/user-attachments/assets/6e61990d-7355-4568-b864-719e2e769345" />

**Document Type**: Mortgage lending rate sheet with multiple table types

### ✅ Good Performance (Rate Tables)
- Simple 3-column rate tables: **95% accuracy**
- LTV/FICO adjustment matrix: **85% accuracy**
- Correct structure recognition and numerical data extraction

### ❌ Poor Performance (Program Highlights Section)
The bottom complex table shows:
- **Structure loss**: Only header rows recognized, data rows missing
- **Text hallucination**: Repeated incorrect tokens (e.g., "NMLY", "INTEREST ONLY")
- **Numerical errors**: "85%" becomes "NMLY", "$1.0MM" becomes "51,000"
- **Nested table confusion**: Cannot handle 3-level nested headers

## Parameters Tested

| Parameter | Default | Tested Values | Result |
|-----------|---------|---------------|--------|
| `base_size` | 1024 | 1536 | Minor improvement (5%) |
| `image_size` | 768 | 1024 | No improvement |
| `crop_mode` | True | False | Slight structure improvement |
| `max_new_tokens` | 8192 | 16384 | No effect (not truncated) |
| Image scaling | 1x | 2x (2804×3502) | **Worse** (forced downscaling) |

## Root Cause Analysis

From debugging output:
```
BASE: torch.Size([1, 256, 1280])
NO PATCHES
```

The problematic table region (~940×280px with ~800 characters) is compressed into **256 visual tokens**, resulting in:
- **Compression ratio**: ~3.1 characters/token
- **Font size**: Original 6-7pt → ~4-5px after base_size scaling
- **Information density**: Exceeds model's visual encoding capacity

## Expected Behavior

The model should:
1. Recognize the multi-level table structure
2. Extract correct LTV/CLTV percentages (e.g., "80%", "85%", "70%")
3. Preserve numerical data like loan amounts ("≤ $1.0MM", "≤ $2.0MM")
4. Handle nested headers properly

## Questions

1. **Training data**: Does the training set include high-density financial tables with small fonts?
2. **Resolution limits**: What's the recommended minimum font size (in pt or px) for reliable recognition?
3. **Workarounds**: Should we:
   - Pre-crop complex table regions separately?
   - Use different `dynamic_preprocess` parameters (min_num/max_num)?
   - Apply specific prompts for financial tables?

## Reproduction Code

```python
from transformers import AutoModel, AutoTokenizer
import torch

model_name = 'deepseek-ai/DeepSeek-OCR-2'
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
model = AutoModel.from_pretrained(
    model_name, 
    _attn_implementation='eager',  # No flash_attn on Windows
    trust_remote_code=True
).eval().cuda().to(torch.bfloat16)

prompt = "<image>\n<|grounding|>Convert the document to markdown."
image_file = 'mortgage_rate_sheet.png'

res = model.infer(
    tokenizer, 
    prompt=prompt, 
    image_file=image_file, 
    output_path='./output',
    base_size=1536,      # Tested with 1024/1536
    image_size=1024,     # Tested with 768/1024
    crop_mode=False,     # Tested True/False
    save_results=True
)
```

## Additional Context

This appears to be a **visual encoding bottleneck** rather than a generation issue, as increasing `max_new_tokens` and adjusting `no_repeat_ngram_size` had no effect. The model stops generating prematurely, suggesting it "believes" recognition is complete despite missing most table content.

Would appreciate any guidance on:
- Model architecture limitations for small-font tables
- Recommended preprocessing strategies
- Whether future versions will improve dense-table handling

---

## Request for Official Review

We've exhaustively tested various parameter combinations and preprocessing strategies, but the issue persists. **Could the team please help verify if we're using the model correctly?** Specifically:

1. **Are we missing any crucial parameters** for handling dense financial tables?
2. **Is there a recommended workflow** for documents with mixed complexity (simple + complex tables)?
3. **Should we be using the vLLM inference path** instead of Transformers for better results?
4. **Are there any undocumented tricks** for improving small-font recognition?

We want to rule out user error before concluding this is a model limitation. Any insights from the team would be greatly appreciated!

**Test artifacts available**: We can provide the full output logs, result.mmd files, and original test images if needed for debugging.

**Comment by bctd7** [comment]
I have also encountered some issues. As a student, when using this tool to recognize complex tables, I found that it repeatedly outputs the same recognized content when processing nested tables. Has your team ever encountered this phenomenon? Thank you for your reply.
